### ISSUE TITLE
Mutate not allowing the same variable to be mutated twice within the same mutate statement

### ISSUE THREAD

**Issue by sahoosatish** [machow/siuba#431]
Hi,
    Please advise if the following is the expected behavior or needs to be incorporated in to  siuba. I tried looking for the error without much luck.

(
comp >>
     mutate(
            gwi = if_else((_.gwi==np.nan) & (_.fyear<=2001), _.GWIE,_.gwi),
            gwi = if_else((_.gwi==np.nan),0,_.gwi)
           )
)
SyntaxError: keyword argument repeated: gwi

of course the following works-

(
comp >>
     mutate(
            gwi = if_else((_.gwi==np.nan) & (_.fyear<=2001), _.GWIE,_.gwi),
                )
)
Is it some Panda-related limitation that Mutate allows the variable to be changed only once per mutate? Thanking you.

**Comment by machow** [comment]
Hey--this is unfortunately a python limitation. Python doesn't allow the same keyword (`gwi=`) to be used twice. 

I think you may also want to use the pandas .isna() method, since there are a range of values that it will flag as NA (None, pd.NA, np.nan), and pandas sometimes converts one NA type to another.

You could use case_when to do multiple things I think...

```
#inside mutate
case_when({
    _.gwi.isna() & (_.fyear <= 2001): _.gwei,
    _.gwi.isna(): 0,
    True: _.gwi
})
```

There is also the pandas `_.gwi.replacena(0)` function if it's useful!

**Comment by sahoosatish** [comment]
Thanks for the quick response Michael.
I thought about the case_when, but not sure if there is any internal ordering to the case_when. I.e. in the above example, I want _.gwi.isna() & (_.fyear <= 2001): _.gwei to be executed first and then whatever is still null I just want to replace that with zero. Since the second condition is a subset of the first, I am worried that if that gets executed first, then the first condition would never be met. I guess what I am saying is case when at least the way it is in SQL, is designed for mutually exclusive cases, which is not the case in my case. Hence, not sure if case_when can rescue me.
I would probably have to do it in two separate steps. I was just looking for a more concise solution, which doesn't seem to exist.
Thanks for the quick response though. Much appreciated.

**Comment by machow** [comment]
It should use the first matching condition, similar to SQL case when. Here's is what it looks like translated to SQL in siuba..!

```python
from siuba.data import cars_sql
from siuba import _, mutate, case_when, show_query
(cars_sql
  >> mutate(
    res = case_when({
        (_.mpg < 10) & (_.cyl == 4): "cond 1",
        (_.mpg > 20) & (_.cyl == 4): "cond 2",
        True: "other"
    }))
  >> show_query()
)
```

```SQL
SELECT
    cars.cyl, cars.mpg, cars.hp,
    CASE 
        WHEN (cars.mpg < 10 AND cars.cyl = 4) THEN 'cond 1'
        WHEN (cars.mpg > 20 AND cars.cyl = 4) THEN 'cond 2'
        ELSE 'other'
    END AS res
FROM cars
```

**Comment by sahoosatish** [comment]
Thanks, Michael. I get that.
But my requirement is something like
(cars_sql
  >> mutate(
    res = case_when({
        (_.mpg < 10) & (_.cyl == 4): "cond 1",
        (_.mpg < 10) : "cond 2",
        True: "other"
    }))
  >> show_query()
)

I.e. conditions 1 and 2 are not mutually exclusive, instead, condition 2 is a subset of condition 1, and only needs to be executed after condition 1 is checked and executed.
Hence, I  think SQL case when (or siuba's implementation of case when) is perhaps not the way to go. Hence, I just chose to do it using two separate mutate statements, one after the other.  It works but is a bit longer to write.But I am ok with that. 
I understand that the way Python pandas and by extension siuba is designed we can't do much here.
Please go ahead and close this thread.
Thanks again for your help.
I am new to siuba but plan to use it extensively. Hope you and the other collaborators keep adding cool features :). Thanks

**Comment by machow** [comment]
Ah, I think I see. Is it that `gwei` can have NAs? In any event I think you can use replacena at the end

```
(
comp >>
     mutate(
            gwi = if_else((_.gwi==np.nan) & (_.fyear<=2001), _.GWIE,_.gwi).replacena(0),
        )
)
```

**Comment by sahoosatish** [comment]
Yes you are correct. GWIE can have NAs. Thanks for the .replacena(0) solution. That would do it for me for this specific case definitely. So I don't have to write two mutates after all :). Thanks.

**Comment by sahoosatish** [comment]
Sorry Michael while implementing .replacena(0) i got the following error. Same with fillna(0) as well. I know these are supported in Panda. Are they supported inside siuba ? Any other syntax that would work here? Thanks

AttributeError: 'numpy.ndarray' object has no attribute 'replacena'

**Comment by machow** [comment]
Ah, sorry--I'm just now realizing that `if_else` and `case_when` return a numpy array. I'll change them to return a Series. (In general, the approach in siuba now is to try and return the same type as the input, e.g. a Series, where possible)

**Comment by machow** [comment]
Alright, should be merged into main! I want to changes to a few other parts of siuba before releasing, but want to get it out in the next week (or at least shortly after RStudioConf)!

**Comment by machow** [comment]
if_else and case_when return Series as of v0.4.0
